sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/StructUtils.java

```diff
@@ -0,0 +1,545 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import com.google.cloud.ByteArray;
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.Type;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.stream.StreamSupport;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.values.Row;
+import org.joda.time.DateTime;
+import org.joda.time.Instant;
+
+final class StructUtils {
+  public static Row translateStructToRow(Struct struct, Schema schema) {
+    checkForSchemasEquality(schema.getFields(), struct.getType().getStructFields(), false);
+
+    List<Schema.Field> fields = schema.getFields();
+    Row.FieldValueBuilder valueBuilder = null;
+    // TODO: Remove this null-checking once nullable fields are supported in cross-language
+    int count = 0;
+    while (valueBuilder == null && count < fields.size()) {
+      valueBuilder = getFirstStructValue(struct, fields.get(count), schema);
+      ++count;
+    }
+    for (int i = count; i < fields.size(); ++i) {
+      valueBuilder = getStructValue(valueBuilder, struct, fields.get(i));
+    }
+    return valueBuilder != null ? valueBuilder.build() : Row.withSchema(schema).build();
+  }
+
+  public static Struct translateRowToStruct(Row row) {
+    Struct.Builder structBuilder = Struct.newBuilder();
+    List<Schema.Field> fields = row.getSchema().getFields();
+    fields.forEach(
+        field -> {
+          String column = field.getName();
+          switch (field.getType().getTypeName()) {
+            case ROW:
+              structBuilder
+                  .set(column)
+                  .to(
+                      beamTypeToSpannerType(field.getType()),
+                      translateRowToStruct(row.getRow(column)));
+              break;
+            case ARRAY:
+              addArrayToStruct(structBuilder, row, field);
+              break;
+            case ITERABLE:
+              addIterableToStruct(structBuilder, row, field);
+              break;
+            case FLOAT:
+              structBuilder.set(column).to(row.getFloat(column).doubleValue());
+              break;
+            case DOUBLE:
+              structBuilder.set(column).to(row.getDouble(column));
+              break;
+            case DECIMAL:
+              structBuilder.set(column).to(row.getDecimal(column).doubleValue());
```

Comment:
This is lossy isn't it? I think we should just refuse to convert DECIMAL since Spanner doesn't have a corresponding type: https://cloud.google.com/spanner/docs/data-types#allowable_types

Reply:
I agree, at first I didn't include decimals but it definitely is lossy.

Reply:
Spanner now supports NUMERIC which may be a better conversion?

> @nielm Could you take a look at this thread? [#12611 (comment)](https://github.com/apache/beam/pull/12611#discussion_r480429441)

Done.

Reply:
Great, I think it wasn't available when I wrote that code. Thanks!